src/Lecture3.hs

```diff
@@ -82,7 +83,25 @@ Tuesday
   would work for **any** enumeration type in Haskell (e.g. 'Bool',
   'Ordering') and not just 'Weekday'?
 -}
-next = error "TODO"
+getTypeValues :: [Weekday]
+getTypeValues = [minBound :: Weekday .. maxBound :: Weekday]
+
+getIndexWithinType :: Weekday -> Maybe Int
+getIndexWithinType v = elemIndex v getTypeValues
+
+correctIndexByMod :: Int -> Int
+correctIndexByMod ind = mod ind typeSize
+  where
+    typeSize = length getTypeValues
+
+next :: Weekday -> Weekday
+next v = go ind
+  where
+    list = getTypeValues
+    ind = getIndexWithinType v
+    go :: Maybe Int -> Weekday
+    go Nothing  = head list -- trap for corner-case
+    go (Just i) = list !! correctIndexByMod (i+1)
```

Comment:
Your implementation is correct 👍🏻 
And you can implement it even shorter by using the standard function `succ`: a method of the `Enum` typeclass.

Reply:
Updated.